from PyQt5 import QtCore, QtGui, QtWidgets
from PyQt5.QtWidgets import QMessageBox, QMainWindow, QStyledItemDelegate, QAction, QLineEdit
import sys

import machine_learning
from source import predict

from PyQt5 import QtCore, QtGui, QtWidgets


class UI_knn(QMainWindow):
    def __init__(self, parent=None):
        super(UI_knn, self).__init__(parent)

    def setupUi(self):
        self.setObjectName("MainWindow")
        self.resize(822, 577)
        self.centralwidget = QtWidgets.QWidget(self)
        self.setWindowIcon(QtGui.QIcon("../icon/logo.png"))
        self.centralwidget.setObjectName("centralwidget")
        self.label_2 = QtWidgets.QLabel(self.centralwidget)
        self.label_2.setGeometry(QtCore.QRect(10, 400, 111, 31))
        font = QtGui.QFont()
        font.setFamily("MS Shell Dlg 2")
        font.setPointSize(10)
        self.label_2.setFont(font)
        self.label_2.setObjectName("label_2")
        self.train = QtWidgets.QPushButton(self.centralwidget)
        self.train.setStyleSheet("background-color:rgb(94, 186, 125)")
        self.train.clicked.connect(self.train_knn)
        self.train.setIcon(QtGui.QIcon('../icon/training.png'))
        self.train.setGeometry(QtCore.QRect(180, 480, 111, 41))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.train.setFont(font)
        self.train.setObjectName("train")
        self.label_15 = QtWidgets.QLabel(self.centralwidget)
        self.label_15.setGeometry(QtCore.QRect(350, 120, 201, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_15.setFont(font)
        self.label_15.setObjectName("label_15")
        self.mae = QtWidgets.QLabel(self.centralwidget)
        self.mae.setStyleSheet("color:red")
        self.mae.setGeometry(QtCore.QRect(540, 90, 251, 16))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.mae.setFont(font)
        self.mae.setObjectName("mae")
        self.visualize = QtWidgets.QPushButton(self.centralwidget)
        self.visualize.setGeometry(QtCore.QRect(650, 360, 131, 31))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.visualize.setFont(font)
        self.visualize.setObjectName("visualize")
        self.label_24 = QtWidgets.QLabel(self.centralwidget)
        self.label_24.setGeometry(QtCore.QRect(350, 180, 151, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_24.setFont(font)
        self.label_24.setObjectName("label_24")
        self.label_13 = QtWidgets.QLabel(self.centralwidget)
        self.label_13.setGeometry(QtCore.QRect(350, 90, 201, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_13.setFont(font)
        self.label_13.setObjectName("label_13")
        self.mse = QtWidgets.QLabel(self.centralwidget)
        self.mse.setStyleSheet("color:red")
        self.mse.setGeometry(QtCore.QRect(540, 120, 251, 16))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.mse.setFont(font)
        self.mse.setObjectName("mse")
        self.label_14 = QtWidgets.QLabel(self.centralwidget)
        self.label_14.setGeometry(QtCore.QRect(350, 150, 201, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_14.setFont(font)
        self.label_14.setObjectName("label_14")
        self.rmse = QtWidgets.QLabel(self.centralwidget)
        self.rmse.setGeometry(QtCore.QRect(540, 150, 251, 16))
        self.rmse.setStyleSheet("color:red")
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.rmse.setFont(font)
        self.rmse.setObjectName("rmse")
        self.X_combo = QtWidgets.QComboBox(self.centralwidget)
        self.X_combo.setGeometry(QtCore.QRect(550, 340, 79, 23))
        self.X_combo.setObjectName("X_combo")
        self.columns = QtWidgets.QListWidget(self.centralwidget)
        self.columns.setGeometry(QtCore.QRect(10, 70, 271, 191))
        self.columns.setObjectName("columns")
        self.train_size = QtWidgets.QLabel(self.centralwidget)
        self.train_size.setGeometry(QtCore.QRect(120, 340, 181, 16))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.train_size.setFont(font)
        self.train_size.setObjectName("train_size")
        self.test_size = QtWidgets.QLabel(self.centralwidget)
        self.test_size.setGeometry(QtCore.QRect(120, 370, 181, 16))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.test_size.setFont(font)
        self.test_size.setObjectName("test_size")
        self.label_10 = QtWidgets.QLabel(self.centralwidget)
        self.label_10.setGeometry(QtCore.QRect(10, 311, 91, 20))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_10.setFont(font)
        self.label_10.setObjectName("label_10")
        self.Y_combo = QtWidgets.QComboBox(self.centralwidget)
        self.Y_combo.setGeometry(QtCore.QRect(550, 380, 79, 23))
        self.Y_combo.setObjectName("Y_combo")
        self.label_9 = QtWidgets.QLabel(self.centralwidget)
        self.label_9.setGeometry(QtCore.QRect(10, 370, 91, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_9.setFont(font)
        self.label_9.setObjectName("label_9")
        self.neighbours = QtWidgets.QLineEdit(self.centralwidget)
        self.neighbours.setGeometry(QtCore.QRect(120, 400, 141, 23))
        self.neighbours.setText("")
        self.neighbours.setCursorPosition(0)
        self.neighbours.setObjectName("neighbours")
        self.test_data = QtWidgets.QLineEdit(self.centralwidget)
        self.test_data.setGeometry(QtCore.QRect(120, 310, 101, 23))
        self.test_data.setObjectName("test_data")
        self.test_size_btn = QtWidgets.QPushButton(self.centralwidget)
        self.test_size_btn.setStyleSheet("background-color:rgb(94, 186, 125)")
        self.test_size_btn.clicked.connect(self.set_size)
        self.test_size_btn.setGeometry(QtCore.QRect(230, 310, 71, 23))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.test_size_btn.setFont(font)
        self.test_size_btn.setObjectName("test_size_btn")
        self.accuracy = QtWidgets.QLabel(self.centralwidget)
        self.accuracy.setStyleSheet("color:red")
        self.accuracy.setGeometry(QtCore.QRect(540, 180, 251, 16))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.accuracy.setFont(font)
        self.accuracy.setObjectName("accuracy")
        self.label_12 = QtWidgets.QLabel(self.centralwidget)
        self.label_12.setGeometry(QtCore.QRect(10, 340, 121, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_12.setFont(font)
        self.label_12.setObjectName("label_12")
        self.conf_mat = QtWidgets.QPushButton(self.centralwidget)
        self.conf_mat.clicked.connect(self.con_matrix)
        self.conf_mat.setEnabled(False)
        self.conf_mat.setGeometry(QtCore.QRect(350, 450, 161, 31))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.conf_mat.setFont(font)
        self.conf_mat.setObjectName("conf_mat")
        self.label_11 = QtWidgets.QLabel(self.centralwidget)
        self.label_11.setGeometry(QtCore.QRect(10, 280, 91, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_11.setFont(font)
        self.label_11.setObjectName("label_11")
        self.line_2 = QtWidgets.QFrame(self.centralwidget)
        self.line_2.setGeometry(QtCore.QRect(320, 60, 20, 581))
        self.line_2.setFrameShape(QtWidgets.QFrame.VLine)
        self.line_2.setFrameShadow(QtWidgets.QFrame.Sunken)
        self.line_2.setObjectName("line_2")
        self.report = QtWidgets.QPlainTextEdit(self.centralwidget)
        font = QtGui.QFont()
        font.setFamily("MS Shell Dlg 2")
        font.setPointSize(11)
        self.report.setFont(font)
        self.report.setGeometry(QtCore.QRect(350, 260, 451, 181))
        self.report.setObjectName("report")
        self.label_25 = QtWidgets.QLabel(self.centralwidget)
        self.label_25.setGeometry(QtCore.QRect(350, 230, 201, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.label_25.setFont(font)
        self.label_25.setObjectName("label_25")
        self.label_19 = QtWidgets.QLabel(self.centralwidget)
        self.label_19.setGeometry(QtCore.QRect(150, 10, 501, 51))
        font = QtGui.QFont()
        font.setFamily("Segoe UI Light")
        font.setPointSize(22)
        font.setBold(False)
        font.setItalic(False)
        font.setWeight(50)
        self.label_19.setFont(font)
        self.label_19.setStyleSheet("color: rgb(11, 34, 57);")
        self.label_19.setObjectName("label_19")
        self.target_2 = QtWidgets.QLabel(self.centralwidget)
        self.target_2.setGeometry(QtCore.QRect(120, 280, 181, 21))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.target_2.setFont(font)
        self.target_2.setObjectName("target_2")
        self.predict = QtWidgets.QPushButton(self.centralwidget)
        self.predict.setStyleSheet("background-color:rgb(94, 186, 125)")
        self.predict.clicked.connect(self.predict_open)
        self.predict.setIcon(QtGui.QIcon('../icon/prediction.png'))
        self.predict.setEnabled(False)
        self.predict.setGeometry(QtCore.QRect(660, 480, 111, 41))
        font = QtGui.QFont()
        font.setFamily("Ubuntu Condensed")
        font.setPointSize(10)
        self.predict.setFont(font)
        self.predict.setObjectName("predict")
        self.label_link = QtWidgets.QLabel(self.centralwidget)
        self.label_link.setGeometry(QtCore.QRect(650, 40, 55, 16))
        self.label_link.hide()
        self.label_link.setObjectName("label_link")
        self.setCentralWidget(self.centralwidget)
        self.menubar = QtWidgets.QMenuBar(self)
        self.menubar.setGeometry(QtCore.QRect(0, 0, 822, 25))
        self.menubar.setObjectName("menubar")
        self.setMenuBar(self.menubar)
        self.statusbar = QtWidgets.QStatusBar(self)
        self.statusbar.setObjectName("statusbar")
        self.setStatusBar(self.statusbar)

        self.retranslateUi()
        # QtCore.QMetaObject.connectSlotsByName(MainWindow)

    def retranslateUi(self):
        _translate = QtCore.QCoreApplication.translate
        self.setWindowTitle(_translate("MainWindow", "MainWindow"))
        self.label_2.setText(_translate("MainWindow", "n_neighbours:"))
        self.train.setText(_translate("MainWindow", "Train"))
        self.label_15.setText(_translate("MainWindow", "Mean Square Error:"))
        self.mae.setText(_translate("MainWindow", ""))
        self.visualize.setText(_translate("MainWindow", "Visualize Boundary"))
        self.label_24.setText(_translate("MainWindow", "Accuracy Score:"))
        self.label_13.setText(_translate("MainWindow", "Mean Absolute Error:"))
        self.mse.setText(_translate("MainWindow", ""))
        self.label_14.setText(_translate("MainWindow", "Root Mean Sq. Error:"))
        self.rmse.setText(_translate("MainWindow", ""))
        self.train_size.setText(_translate("MainWindow", ""))
        self.test_size.setText(_translate("MainWindow", ""))
        self.label_10.setText(_translate("MainWindow", "Train Size:"))
        self.label_9.setText(_translate("MainWindow", "Test Size:"))
        self.neighbours.setPlaceholderText(_translate("MainWindow", "Number of neighbors"))
        self.test_data.setText(_translate("MainWindow", "0.1"))
        self.test_size_btn.setText(_translate("MainWindow", "Set"))
        self.accuracy.setText(_translate("MainWindow", ""))
        self.label_12.setText(_translate("MainWindow", "Train Size:"))
        self.conf_mat.setText(_translate("MainWindow", "Confusion Matrix"))
        self.label_11.setText(_translate("MainWindow", "Target : "))
        self.label_25.setText(_translate("MainWindow", "Classification Report:"))
        self.label_19.setText(_translate("MainWindow", "K-Nearest Neighbors Classifier"))
        self.target_2.setText(
            _translate("MainWindow", "<html><head/><body><p>Dropout or Graduate<br/></p></body></html>"))
        self.predict.setText(_translate("MainWindow", "Predict"))
        self.label_link.setText(_translate("MainWindow", "TextLabel"))

    def set_size(self):
        try:
            path = self.label_link.text()
            number = self.test_data.text()
            size = machine_learning.size_split(path, float(number))
            self.train_size.setText(str(size[0]))
            self.test_size.setText(str(size[1]))
        except:
            QMessageBox.information(self, 'System',
                                    'An error occurred while setting test, train size. Please try again!',
                                    QMessageBox.Close)

    def train_knn(self):
        try:
            number = self.test_data.text()
            path = self.label_link.text()
            n_neigh = self.neighbours.text()
            result = machine_learning.knn(path, 'error', float(number), int(n_neigh))
            self.mae.setText(str(result[0]))
            self.mse.setText(str(result[1]))
            self.rmse.setText(str(result[2]))
            self.accuracy.setText(str(result[3]))
            self.report.setPlainText(str(result[4]))
        except:
            QMessageBox.information(self, 'System', 'An error occurred while training. Please try again!',
                                    QMessageBox.Close)
        else:
            self.predict.setEnabled(True)
            self.conf_mat.setEnabled(True)

    def con_matrix(self):
        try:
            number = self.test_data.text()
            path = self.label_link.text()
            n_neigh = self.neighbours.text()
            result = machine_learning.knn(path, 'matrix', float(number), int(n_neigh))
        except:
            QMessageBox.information(self, 'System',
                                    'An error occurred while displaying the confusion matrix. Please try again!',
                                    QMessageBox.Close)

    def predict_open(self):
        try:
            n_neigh = self.neighbours.text()
            number = self.test_data.text()
            self.ui = QMainWindow()
            self.ui = predict.Ui_predict_window(n_neigh,number)
            self.ui.setupUi()
            self.ui.label_link.setText(self.label_link.text())
            self.ui.label_tt.setText('K-Nearest Neighbors Classifier')
            self.ui.show()
        except:
            QMessageBox.information(self, 'System',
                                    'An error occurred while opening predict. Please try again!',
                                    QMessageBox.Close)


if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)
    ui = UI_knn()
    ui.setupUi()
    ui.show()
    sys.exit(app.exec_())
